sql/sql_repl.cc

```diff
@@ -2232,15 +2239,118 @@ send_event_to_slave(binlog_send_info *info, Log_event_type event_type,
-      return NULL;
+    const uchar *buf = (const uchar*) packet->ptr() + ev_offset;
+
+      Log_event_type etype = (Log_event_type) buf[EVENT_TYPE_OFFSET];
+
+      // Only ROW events have m_flags
+      bool is_rows_event =
+          (etype == WRITE_ROWS_EVENT)        ||
+          (etype == UPDATE_ROWS_EVENT)       ||
+          (etype == DELETE_ROWS_EVENT)       ||
+          (etype == WRITE_ROWS_EVENT_V1)     ||
+          (etype == UPDATE_ROWS_EVENT_V1)    ||
+          (etype == DELETE_ROWS_EVENT_V1);
+      bool is_gtid_event = (etype == GTID_EVENT);
+
+    if (is_gtid_event)
+    {
+      info->row_events_sent_before= false; // reset per event group
+    }
+    
+    if ((event_flags & LOG_EVENT_SKIP_REPLICATION_F))
+    {
+      /*
+        if the current event has the STMT_F flag as well as the
+        LOG_EVENT_SKIP_REPLICATION_F flag then we should clear the buffer
+        data of this event and rewrite this *_ROW_EVENT to avoid sending
+        filtered data to the replica otherwise if it does not have
+        the STMT_F flag then it should just get filterd 
+        (return NULL basically filters and does not send the event
+        over the network)
+      */
+
+      if (!is_rows_event) {
+          return NULL;
+      }
+
+      /*
+        Extract the STMT_END_F of the rows_event flag
+      */
+      uint8 common_header_len = info->fdev->common_header_len;
+
+      uint8 post_header_len = info->fdev->post_header_len[etype - 1];
+
+      const uchar *post_start = buf + common_header_len;
+
+      post_start += RW_MAPID_OFFSET;
+
+      uint64 table_id;
+
+      if (post_header_len == 6)
+      {
+          table_id= uint4korr(post_start);
+          post_start += 4;
+      }
+      else
+      {
+        table_id  = uint4korr(post_start);
+        table_id |= ((uint64) uint2korr(post_start + 4)) << 32;
+        // skip 6 bytes
+        post_start += RW_FLAGS_OFFSET;
+      }
+
+      uint16 m_flags = uint2korr(post_start);
+
+      bool has_stmt_end_f = (m_flags & Rows_log_event::STMT_END_F);
+
+
+      if (!has_stmt_end_f)
+      {
+        return NULL;
+      }
+      else {
+        if (!info->row_events_sent_before) {
+          return NULL;
+        }
+
+        /* 
+          TODO: Clear payload and rewrite the checksum
+        */
+        // const uchar *row_data_start = buf + common_header_len + post_header_len;
+        // size_t checksum_len= (info->current_checksum_alg 
+        //                       != BINLOG_CHECKSUM_ALG_OFF && 
+        //                       info->current_checksum_alg 
+        //                       != BINLOG_CHECKSUM_ALG_UNDEF) 
+        //               ? BINLOG_CHECKSUM_LEN : 0;
+        // size_t row_data_len = len - (row_data_start - buf) - checksum_len;
+
+        // memset((uchar*)row_data_start, 0, row_data_len); // zero out row payload
+        // if (checksum_len > 0) 
+        // {
+        //   uint32 crc = my_checksum(0, buf,  len - BINLOG_CHECKSUM_LEN);
```

Comment:
Hmm, I'm not sure if it makes a difference, but in the code, the header checksum happens separately from the data checksum. So if you're running into a problem with this, it might make more sense to do:

```
uint32 crc= my_checksum(0, buf, common_header_len);
crc= my_checksum(crc, buf + common_header_len, len - common_header_len - BINLOG_CHECKSUM_LEN);
```